[Security Solution] Incorporates EQL options in EQL query validation on both Rule Creation and Timeline

rylnd opened the pull request:
## Summary

This PR updates the Detection Rule Creation and Timeline forms to account for the new(er) EQL options fields (`timestamp_field`, `event_category_field`, and `tiebreaker_field`) when validating the EQL query. While the rule query and timeline query (respectively) would correctly persist and use these options, they were unused during EQL validation, meaning that in certain situations it was impossible to produce a "valid" (from the perspective of the form) EQL query if your data necessitated one of those options (see https://github.com/elastic/kibana/issues/158326 for more details).

### Changes
The above was accomplished by:

1. Adding a hidden form field to the UI for `eqlOptions`
    * A lack of a field component meant that `formData` never held any value for this field
    * This meant that the form-based validations (i.e. `eqlValidator`) would not have access to `eqlOptions`
2. Updating the `onOptionsChange` hook to additionally inform the form of changes to `eqlOptions`
    * This allows the form to have updated values for `eqlOptions` with minimal changes
    * Not ideal from a data-flow perspective, since the field changes are updating two copies of `eqlOptions`. However, we're only using the "form copy" for the validation call, and everything else works as it did before.

 
### Additional changes
__Update__: I've reverted the described changes to the default timeline values below, and moved that to a [separate issue](https://github.com/elastic/security-team/issues/8892).

In the course of performing the above tasks for timeline, I found the behavior there to be slightly different: we were specifying default values for these fields, which was redundant in some cases (`timestamp_field`, `event_category_field`), and invalid in the case of `tiebreaker_field`. You can see here how the `tiebreaker_field` must be changed to make the query valid: 

https://github.com/elastic/kibana/assets/657252/556bc998-c147-4825-8bde-5d8d03873e75


~~I opted to remove these default values, as well as the defaulting behavior from the search strategy call (see 10a47fdaba5860a44638aada78c0cf8c37b2c580). While these values may be persisted in existing saved timelines, this makes the EQL workflow much cleaner for new users, and those users with persisted EQL options can now modify and validate those options on the fly.~~

#### Steps to Review

For @elastic/kibana-visualizations and/or @elastic/kibana-data-discovery : I've added a few tests to the EQL search strategy that document existing behavior RE EQL option fields. Nothing about the search strategy itself was changed. If you think the tests aren't useful, I'm happy to delete them.

1. Create an index without a `@timestamp` field, and insert some data:
    <details>
    <summary>Dev Tools</summary>

       PUT timestamp/
       {
         "mappings": {
           "dynamic": "strict",
           "properties": {
             "timestamp": {
               "type": "date",
               "format": "epoch_second"
             },
             "locale": {
               "type": "keyword"
             },
             "created_at": {
               "type": "date",
               "format": "epoch_second"
             },
             "event": {
               "properties": {
                 "category": {
                   "ignore_above": 1024,
                   "type": "keyword"
                 }
               }
             }
           }
         }
       }

       PUT timestamp/_doc/4
       {
         "timestamp": 1710211630,
         "locale": "es",
         "created_at": 1710211630,
         "event.category": "configuration"
       }
    </details>
2. Create a new EQL detection rule, and choose `timestamp` as the index pattern
3. Type `any where true` as the query, and observe a validation error (`Found 1 problem line -1:-1: Unknown column [@timestamp]`)
4. Open EQL Settings popover, and select `timestamp` from the `Timestamp field` dropdown
5. Observe that the query has been revalidated, and found valid.


### New Behavior:

https://github.com/elastic/kibana/assets/657252/22a63363-d597-4dfd-8f9a-647f4084ec0e



### Related Issues
* Addresses https://github.com/elastic/kibana/issues/158326

### Checklist

Delete any items that are not applicable to this PR.

- [ ] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [x] [Flaky Test Runner](https://ci-stats.kibana.dev/trigger_flaky_test_runner/1) was used on any tests changed

### For maintainers

- [x] This was checked for breaking API changes and was [labeled appropriately](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)


<!--ONMERGE {"backportTargets":["8.13"]} ONMERGE-->

rylnd committed:
Document current EQL rule behavior with timestamp fields

The conclusion is that we currently need _both_ of these fields for the
EQL rule to operate if the data doesn't have a @timestamp at all.
However, the UI does not allow you to set both of these fields.

rylnd committed:
Add similar tests for data with @timestamp

rylnd committed:
Add tests validating that our EQL search strategy respects certain parameters

This is a sanity check to allowing our EQL validation functionality to
leverage these same options.

rylnd committed:
EQL validation function accepts EQL options as a parameter

This will break our calls, which is what we want (they need to pass
options now).

rylnd committed:
Makes eqlOptions a used field on our Define Step form

Previously, the eqlOptions field was referenced in the schema of the
Define Step form, but as there was no component controlling the field,
it was unknown to the form data, and consequently unavailable to the
queryBar validator (which is ultimately our goal here).

By adding a component for the eqlOptions field (albeit hidden), and
updating our onOptionsChange callback to additionally set that field
data, the rule form now has eqlOptions available during validation of
the EQL query.

rylnd committed:
Ensure that changes to EQL options cause the EQL query to be revalidated

This is one case where the form lib makes that super easy 👍.

rylnd committed:
Add a test validating how onOptionsChange is called

This was mostly a sanity check, but good to document this API IMO.

rylnd committed:
Remove unused interface

rylnd committed:
Add an eqlOptions form field to the timeline EQL form

By adding the field to the form itself (as a hidden field), and
modifying the onOptionsChange callback to additionally update the form
value, we now have that field value available to us when validating the
EQL query itself.

This is all near-identical to the changes just made to
the Define Step rule creation form.

rylnd committed:
Remove default values from timeline EQL options

* The default value of `''` for `tiebreaker_field` is invalid and must be
  changed by the user
* The redux store selector was using default values that both didn't make sense and
  were never used
* Removes defaulting behavior for options in the timeline EQL search
  strategy, since sending `timestamp_field: undefined` is equivalent to
  not specifying any value (and thus falls back to the equivalent EQL
  default)

rylnd committed:
Merge branch 'main' into rylnd/bugfix/eql_rule_timestamp_override

rylnd commented:
/ci

rylnd committed:
Rename new archive to be more accurate

While the data I wrote was "non-ECS," that's not a necessary feature of
the bug here. It simply needs to _not_ have a @timestamp field.

This also updates the corresponding FTR tests to not be exclusive
(whoops).

rylnd committed:
Rename archive, again

While the '@' in the file/folder seems to work, I'm a little dubious of
the portability. Using normal characters here just to be safe.

rylnd committed:
More accurate cypress selector

This is retrieving the general rule index pattern input (of which there
is no other reference); I think the naming was due to it originally
being introduced while testing the IM rule type.

rylnd commented:
/ci

rylnd committed:
Add a cypress test around the EQL option/validation functionality

This automates the process that I've been using to test this: given data
without an `@timestamp` (the default time field for EQL), you'll see
errors when we attempt to validate the EQL query UNTIL you choose
another timestamp field via the EQL options.

rylnd commented:
/ci

elasticmachine commented:
Pinging @elastic/security-detection-engine (Team:Detection Engine)

rylnd commented:
>  I’ve seen very few instances where a user needed to not use ‘@timestamp’ 

@kqualters-elastic I mentioned it in https://github.com/elastic/kibana/commit/10a47fdaba5860a44638aada78c0cf8c37b2c580, but specifying `timestamp_field: '@timestamp'` is equivalent to not specifying it at all, since `@timestamp` is the default value (at the [EQL level](https://www.elastic.co/guide/en/elasticsearch/reference/current/eql.html#eql-required-fields)).

As I see it, the two arguments for keeping the defaulting behavior are:
1. There are users that have experienced this behavior, and likely now have saved timelines with those values
2. It's educational to show the user what the language-level defaults are for those fields.

This behavior is already inconsistent between timeline and rule creation: only timeline has defaults for these fields.

rylnd committed:
Prevent component props from being passed as DOM attributes

When you don't specify a component with UseField, it effectively renders
something like `() => <input {...props} />`, which in this case meant
that `euiFieldProps` was being set as a DOM attribute on the input
element, and react was complaining about that.

This solves the issue by specifying a component to use (HiddenField,
which is effectively `() => null`), and consequently removes the
unnecessary `display: none` wrapper as well.

Since this field is never displayed, one wonders why it has those
`euiFieldProps` at all, but that's a question for another day.

semd commented on x-pack/plugins/security_solution/public/common/hooks/eql/api.ts:
```diff
@@ -23,6 +24,7 @@ interface Params {
   data: DataPublicPluginStart;
   signal: AbortSignal;
   runtimeMappings: estypes.MappingRuntimeFields | undefined;
+  options: Omit<EqlOptionsSelected, 'query' | 'size'>;
```
Should this one be optional?
```suggestion
  options?: Omit<EqlOptionsSelected, 'query' | 'size'>;
```

rylnd replied:
I made it required to begin in order to light up all the call sites with typescript. As with `runtimeMappings` I tend to prefer arguments to be required (but allow `undefined` if it makes sense) so I'd likely do the same here (although `options: {}` also satisfies). Since there's just the single call currently, I'm leaning toward explicitness over developer ergonomics.

semd replied:
It makes complete sense, but since it is called "_options_", it denotes they are values that can be optionally defined, or just omitted (that's not the case for `runtimeMappings`). So, in this specific case, I would lean towards making it optional instead of forcing the "options" to be always specified.

Nevertheless, that's just a subtility (pretty subjective probably), so I think we are perfectly good anyway with `options: Omit<EqlOptionsSelected, 'query' | 'size'> | undefined;` 👍

rylnd replied:
Ahh, yes, I see your point now! "Options" does imply "optionality" 😉

rylnd replied:
e837f19373ff9f7b99dcc6825878f62b055afc1a

rylnd commented:
@kqualters-elastic RE your [earlier comment](https://github.com/elastic/kibana/pull/178468#issuecomment-1993084855): 

The timeline behavior shown in my video above (`Screen.Recording.2024-03-11.at.10.12.28.PM.mov`) is _new_ and was introduced when this PR added those EQL options to the validation call.

The `tiebreaker_field: ''` default worked previously due to [this code](https://github.com/elastic/kibana/blob/main/x-pack/plugins/timelines/server/search_strategy/timeline/eql/helpers.ts#L54-L58) effectively ignoring that value; the use of `isEmpty` leads me to believe that that guard was added for just this default case.

So: the quickest solution here would simply be to add a similar guard on the validation call. I think that's probably warranted either way, given that users may rehydrate saved timelines with `tiebreaker_field: ''` in the future.

However, there are the broader issues of:

1. The `tiebreaker_field` default being invalid
2. The inconsistency with the rule creation form (where we don't specify default values for those EQL options)

that would be nice to resolve, as well.

rylnd committed:
Revert "Remove default values from timeline EQL options"

This reverts commit 10a47fdaba5860a44638aada78c0cf8c37b2c580.

While we're still discussing how to handle the invalid default for
`tiebreaker_field`, and the discrepancy with the rule creation form,
these changes are not immediately necessary to fix the bug at hand.

rylnd committed:
Guard against using timeline default of '' for tibreakerField

An empty string is the default value for the EQL option
`tiebreaker_field` in the timeline EQL form. However, if this is passed
directly to ES an error will be thrown, as this is never a valid value
for that option.

A similar guard exists in the timeline EQL search strategy; we're adding
one to our validation call as well now that it has to deal with those
values.

rylnd committed:
validateEql options parameter is optional

These additional parameters do not need to be specified in order to
produce a valid query, but I am requiring that the key be specified for
sanity's sake.

rylnd commented:
@kqualters-elastic I split off the defaulting conversation into https://github.com/elastic/security-team/issues/8892. On this PR, I've reverted the defaulting changes, and added the necessary guard to prevent the behavior in the above video.

rylnd commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/components/step_define_rule/index.tsx:
```diff
@@ -814,17 +827,16 @@ const StepDefineRuleComponent: FC<StepDefineRuleProps> = ({
     <>
       <StepContentWrapper addPadding={!isUpdateView}>
         <Form form={form} data-test-subj="stepDefineRule">
-          <StyledVisibleContainer isVisible={false}>
```
This is wholly unrelated to this PR, but react was throwing dev warnings on this page and I couldn't help but fix it. See 3a9d04059f9e8bb22ca77aeff15d7028cb7f3d98 for more details.

rylnd commented on x-pack/test/security_solution_cypress/cypress/e2e/detection_response/detection_engine/rule_creation/event_correlation_rule.cy.ts:
```diff
@@ -174,4 +183,34 @@ describe('EQL rules', { tags: ['@ess', '@serverless'] }, () => {
     });
   });
+
+  describe('with source data requiring EQL overrides', () => {
```
This is the one true 🔴 -> 🟢 test in this PR; the rest are just documenting existing behavior.

kqualters-elastic commented:
@rylnd cool, I ran everything locally before that commit and all lgtm 👍 I think an easy solution (and could probably revert to how you had it) would be to just add the default field to either the form field description label or the input placeholder

lukasolson reviewed (approved):
DataDiscovery code changes LGTM!

rylnd committed:
Merge branch 'main' into rylnd/bugfix/eql_rule_timestamp_override

rylnd committed:
Merge branch 'main' into rylnd/bugfix/eql_rule_timestamp_override

rylnd committed:
Remove duplicated es_archive thanks to new cypress task functionality

In #178724, we added the ability to retrieve archives from another
location (namely, where our FTR archives live). This allows me to delete
the duplicated archive specific for the cypress usage, and share the FTR
archive between both. Nice!

vitaliidm commented on x-pack/test/security_solution_api_integration/test_suites/detections_response/detection_engine/rule_execution_logic/trial_license_complete_tier/execution_logic/eql.ts:
```diff
@@ -746,5 +746,140 @@ export default ({ getService }: FtrProviderContext) => {
+
+        expect(_log.errors).to.be.empty();
+        expect(_log.warnings).to.eql([
+          'This rule reached the maximum alert limit for the rule execution. Some alerts were not created.',
```
test title suggests no alerts were generated, but they were
`specifying only timestamp_field results in a warning, and no alerts are generated`

rylnd replied:
Oops! Good catch. f3e55d52877

vitaliidm commented on x-pack/test/security_solution_api_integration/test_suites/detections_response/detection_engine/rule_execution_logic/trial_license_complete_tier/execution_logic/eql.ts:
```diff
@@ -746,5 +746,140 @@ export default ({ getService }: FtrProviderContext) => {
+          logs: [_log],
+        } = await previewRule({ supertest, rule });
+
+        expect(_log.errors).to.be.empty();
```
test title suggests no alerts were generated and error encountered
no error in test and alert were generated
`specifying only timestamp_override results in an error, and no alerts are generated`

rylnd replied:
f3e55d52877

vitaliidm commented on x-pack/test/security_solution_api_integration/test_suites/detections_response/detection_engine/rule_execution_logic/trial_license_complete_tier/execution_logic/eql.ts:
```diff
@@ -746,5 +746,140 @@ export default ({ getService }: FtrProviderContext) => {
+        expect(previewAlerts.length).to.be.greaterThan(0);
+      });
+
+      it('specifying both timestamp_override and timestamp_field behaves as expected', async () => {
```
can we add more details what is expected in this case.

rylnd replied:
Good catch, thank you! f3e55d52877

rylnd committed:
Merge branch 'main' into rylnd/bugfix/eql_rule_timestamp_override

rylnd committed:
Fix type incompatibility with EQL search strategy

The actual ES client/EQL endpoint allows for specifying options at both
the top level _and_ within the request's `body` parameter. However, we
had only typed the latter in our original implementation.

This aligns the types with the actual (demonstrable, as evidenced by all
the @ts-expect-errors) behavior of the server, using a
union type cribbed from responseops (see #171348).

Consequently, all these `@ts-expect-error`s can now be removed.

rylnd committed:
Fix test descriptions to align with test bodies

A few of these were updated in the course of development without the
title being updated. This clarifies the behavior in the case of data
with `@timestamp`, as well as replacing the vague "behaves as expected"
descriptions with more helpful ones.

rylnd commented on x-pack/plugins/security_solution/public/common/hooks/eql/api.ts:
```diff
@@ -31,13 +33,18 @@ export const validateEql = async ({
         params: {
           index: dataViewTitle,
           body: { query, runtime_mappings: runtimeMappings, size: 0 },
+          // @ts-expect-error top-level keys are unexpected in {@link EqlSearchRequest}
```
These `@ts-expect-error`s were due to an incompatibility between the actual EQL endpoint, and the types in the search strategy. I managed to extend the type to make this all work without the need for the `@ts-expect-error`: 676718221eb.

I'm not sure what you were referring to with the `await` comment, though.

rylnd commented on x-pack/test/security_solution_cypress/cypress/e2e/detection_response/detection_engine/rule_creation/event_correlation_rule.cy.ts:
```diff
@@ -174,4 +183,34 @@ describe('EQL rules', { tags: ['@ess', '@serverless'] }, () => {
+      cy.task('esArchiverUnload', { archiveName: 'no_at_timestamp_field', type: 'ftr' });
+    });
+
+    it('includes EQL options in query validation', () => {
```
[Of course](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/5526). 

Do you see anything in the test itself that looks suspect?

vitaliidm replied:
nothing suspicious, but when adding new test we usually run them in flaky test runner to ensure they are not flaky, as an effort to reduce overall Cypress tests flakiness.
Since even perfectly written code does not guarantee tests won't be flaky unfortunately

rylnd commented:
@vitaliidm thanks for the diligence in your review! I fixed the type error, as well as the test descriptions, and the flaky runner is running now. Ready for another review!

vitaliidm reviewed (approved):
Detection Engine area LGTM
@rylnd thank you for fixing it, adding extensive test coverage and addressing my review comments

kibanamachine commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.13|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/178994">](https://github.com/elastic/kibana/pull/178994)|

Note: Successful backport PRs will be merged automatically after passing CI.

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

<!--- Backport version: 9.4.3 -->
